### PR TITLE
Test suite

### PR DESCRIPTION
**Pull request by jchen293**
This PR adds full test coverage to the library by testing everything "reasonable". The test coverage is 78%

Screen Shot 2022-03-15 at 1 23 37 PM

### CODE REVIEW

File: tests/address_test.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+
+@pytest.mark.vcr()
+def test_address_retrieve():
+    address = easypost.Address.create(**fixture.basic_address())
+
+    retrieved_address = easypost.Address.retrieve(address.id)
+
+    assert isinstance(retrieved_address, easypost.Address)
+    assert retrieved_address.id == address.id
```
Comment (jchen293) [inline tests/address_test.py:35]: # reviewer note:

We can't compare instances here because we have to implement the method [__eq__](https://docs.python.org/3/reference/datamodel.html#object.__eq__) in each class

Comment (nwithan8) [reply tests/address_test.py:35]: We could probably do some kind of "loop over every property of Object A, get the corresponding property on Object B, compare", but I think that would introduce some funky stuff. This works for now.

File: tests/batch_test.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+@pytest.mark.vcr()
+def test_batch_create_scanform():
+    shipment_data = fixture.one_call_buy_shipment()
+
+    batch = easypost.Batch.create(shipments=[shipment_data])
+    batch.buy()
+
+    # Uncomment the following line if you need to re-record the cassette
+    time.sleep(5)  # Wait enough time for the batch to process buying the shipment
```
Comment (nwithan8) [inline tests/batch_test.py:69]: If we've recorded the cassette already, we'll want to comment this out.

File: tests/batch_test.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+@pytest.mark.vcr()
+def test_batch_label():
+    shipment_data = fixture.one_call_buy_shipment()
+
+    batch = easypost.Batch.create(shipments=[shipment_data])
+    batch.buy()
+
+    # Uncomment the following line if you need to re-record the cassette
+    time.sleep(5)  # Wait enough time for the batch to process buying the shipment
```
Comment (nwithan8) [inline tests/batch_test.py:100]: Comment this out if we've recorded already

File: tests/carrier_account_test.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            "account_number": "A1A1A1",
+            "user_id": "USERID",
+            "password": "PASSWORD",
+            "access_license_number": "ALN",
+        },
+    }
+
+
+@pytest.mark.vcr()
+def test_carrier_account_create(prod_api_key):
```
Comment (nwithan8) [inline tests/carrier_account_test.py:20]: Is prod key actually being used in here? Check for all these tests

**Review by nwithan8 — CHANGES_REQUESTED** [review]

File: tests/batch_test.py (hunk 2 above)
Comment (jchen293) [reply tests/batch_test.py:69]: oops, nice catch!

File: tests/carrier_account_test.py (hunk 4 above)
Comment (jchen293) [reply tests/carrier_account_test.py:20]: Yes, the default API key is using the test key, we need to pass in prod_api_key for carrier account testing since it require prod key.

Comment (nwithan8) [reply tests/carrier_account_test.py:20]: What I'm saying is, it doesn't seem like the prod key is actually getting passed along to the inner functions to actually be used? Or am I missing some syntactical sugar?

Comment (jchen293) [reply tests/carrier_account_test.py:20]: `prod_api_key` is coming from `conftest.py` as a fixture. Check out this doc: https://docs.pytest.org/en/latest/how-to/fixtures.html

**Review by nwithan8 — DISMISSED** [review]

File: tests/address_test.py (hunk 1 above)
Comment (Justintime50) [reply tests/address_test.py:35]: I played briefly with this. Only comparing IDs isn't ideal when we can reliably compare the entire object (various bugs exposed themselves due to comparing whole objects in other libs). I also implemented `__eq__` in a class to try and didn't have luck, though I may have done so incorrectly on my first attempt.

My suggestion is we try to compare whole objects. I'd suggest doing a bit of extra digging here to see what our options are and give it a try on a couple objects.

File: easypost/requestor.py [hunk 5]
```diff
@@ -117,9 +117,10 @@ def requests_request(self, method, abs_url, headers, params):
             http_body = result.text
             http_status = result.status_code
         except Exception as e:
+            breakpoint()
```
Comment (Justintime50) [inline easypost/requestor.py:120]: Please double check your own code prior to pushing it (self-review each file before you commit in your IDE), stuff like this should never make it into the repo.

File: easypost/requestor.py [hunk 6]
```diff
@@ -117,9 +117,10 @@ def requests_request(self, method, abs_url, headers, params):
...
             raise Error(
                 "Unexpected error communicating with EasyPost. If this "
-                f"problem persists please let us know at {SUPPORT_EMAIL}.",
+                f"problem persists please let us know at {SUPPORT_EMAIL}...",
```
Comment (Justintime50) [inline easypost/requestor.py:123]: Let's leave this as it was.

File: tests/insurance_test.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+@pytest.mark.vcr()
+def test_insurance_create():
+    shipment = easypost.Shipment.create(**fixture.one_call_buy_shipment())
+
+    insurance = easypost.Insurance.create(
+        to_address=fixture.basic_address(),
+        from_address=fixture.basic_address(),
+        tracking_code=shipment.tracking_code,
+        carrier=fixture.usps(),
+        amount="100",
```
Comment (Justintime50) [inline tests/insurance_test.py:16]: Nit: You could use a fixture for this instead (see the Node PR for an example)

File: tests/insurance_test.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+@pytest.mark.vcr()
+def test_insurance_retrieve():
+    shipment = easypost.Shipment.create(**fixture.one_call_buy_shipment())
+
+    insurance = easypost.Insurance.create(
+        to_address=fixture.basic_address(),
+        from_address=fixture.basic_address(),
+        tracking_code=shipment.tracking_code,
+        carrier=fixture.usps(),
+        amount="100",
```
Comment (Justintime50) [inline tests/insurance_test.py:33]: Nit: You could use a fixture for this instead (see the Node PR for an example)

File: tests/refund_test.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import tests.fixture as fixture
+
+
+@pytest.mark.vcr()
+def test_refund_create():
+    shipment = easypost.Shipment.create(**fixture.one_call_buy_shipment())
+
+    retrieved_shipment = easypost.Shipment.retrieve(
+        shipment
+    )  # We need to retrieve the shipment so that the tracking_code has time to populate
```
Comment (Justintime50) [inline tests/refund_test.py:13]: Comments like this may read better if they came just before the code (that way they aren't formatted weird due to `Black`)

File: tests/report_test.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    except Exception:
+        assert Exception
+
+
+@pytest.mark.vcr()
+def test_report_all_no_type():
+    try:
+        _ = easypost.Report.all()
+    except Exception:
+        assert Exception
```
Comment (Justintime50) [inline tests/report_test.py:134]: These tests aren't necessary. They were included in PHP for example because PHP had specific branching logic to that should have been tested whereas the Python lib does not have that.

Delete the cassettes along with this so they don't stay in the repo.

File: tests/shipment_test.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+@pytest.mark.vcr()
+def test_shipment_convert_label():
+    shipment = easypost.Shipment.create(**fixture.full_shipment())
+    shipment.buy(rate=shipment.lowest_rate())
+    shipment.label(file_format="ZPL")
+
+    assert shipment.postage_label.label_zpl_url is not None
+
+
+# If the shipment was purchased with a USPS rate, it must have had its insurance set to `0` when bought
```
Comment (Justintime50) [inline tests/shipment_test.py:69]: In Python, these large block comments should instead be docblocks, please move them just below the test function name and wrap them in trouble quotes. This is true for all the test comments that are at the "function" level.

**Review by Justintime50 — CHANGES_REQUESTED** [review]
- The convention in Python is to name test files as `test_xyz`, you have them the opposite order. Please rename all the test files so that they start with the word `test` followed by an underscore and then the name of the object being tested.

- Please also replace the two occurrences of `@pytest.yield_fixture` in the `conftest.py` file with `@pytest.fixture` as this is now deprecated and throwing warnings on test runs.

- You can also remove the `per_run_unique` fixture from `conftest.py`, I don't think we need/use it anymore

File: tests/easypost_test.py [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import easypost.requestor as requestor
+
+
+def test_timeout():
+    assert requestor.timeout == 60
+
+    requestor.timeout = 1
```
[the remark this answers is not in the document]
Comment (jchen293) [reply tests/easypost_test.py:7]: I just did, this is resolved.

File: tests/address_test.py (hunk 1 above)
Comment (jchen293) [reply tests/address_test.py:35]: I have created a helper function that is going to compare each attribute recursively. We do have to convert the instance to dictionary first for the comparsion.

File: tests/user_test.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+def test_user_retrieve(prod_api_key):
+    user = easypost.User(fixture.child_user_id())
+
+    assert isinstance(user, easypost.User)
+    assert str.startswith(user.id, "user_")
+
+
+@pytest.mark.vcr()
+def test_user_retrieve_me(prod_api_key):
+    user = easypost.User(fixture.child_user_id())
```
[the remark this answers is not in the document]
Comment (jchen293) [reply tests/user_test.py:27]: Oops, I forgot to change it after copy-paste.

File: tests/carrier_account_test.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    assert carrier_account.description == test_description
+
+
+@pytest.mark.vcr()
+def test_carrier_account_delete(prod_api_key):
+    carrier_account = easypost.CarrierAccount.create(**generate_carrier_account())
+
+    response = carrier_account.delete()
+
+    assert response is not None
```
[the remark this answers is not in the document]
Comment (jchen293) [reply tests/carrier_account_test.py:68]: I think Python lib is a little different, the `delete()` method is going to refresh the object and does not return anything from the endpoint. I will just remove this assertion.

[the remark this answers is not in the document]
Comment (Justintime50) [reply tests/carrier_account_test.py:68]: Erm... if the object gets deleted then it can't return an object haha...

[the remark this answers is not in the document]
Comment (jchen293) [reply tests/carrier_account_test.py:68]: I just realized there is some issue with `refresh_from()` it's being called from the `delete()` method. The object is actually never updated to the empty response. Looking into this now.

[the remark this answers is not in the document]
Comment (jchen293) [reply tests/carrier_account_test.py:68]: Adding an if statement to check if the response is empty in the `delete()` method.

File: tests/insurance_test.py (hunk 7 above)
Comment (jchen293) [reply tests/insurance_test.py:16]: A helper function is created on the top of test_insurance.py

File: tests/report_test.py (hunk 10 above)
Comment (jchen293) [reply tests/report_test.py:134]: removed

File: tests/address_test.py (hunk 1 above)
Comment (jchen293) [reply tests/address_test.py:35]: My own customized helper function didn't work well because the response object can have different types of attributes. Spent a few hours trying to make my helper functions work but later realized we can use [deepdiff](https://pypi.org/project/deepdiff/) to compare two instances if that's okay?

Comment (Justintime50) [reply tests/address_test.py]: I'd rather not introduce another dependency if we can get away without it. Thinking on this a bit, I believe I may have seen an implementation of something like this from @jontsai in the past (that we use internally) that may not even be that many lines of code that we could snag and add here. We can ask him about this.

File: tests/test_pickup.py [hunk 15]
```diff
@@ -1,148 +1,56 @@
...
-import time
-
 import pytest
-import pytz
 
 import easypost
 
-ONE_DAY = datetime.timedelta(days=1)
 
+def generate_pickup(one_call_buy_shipment, basic_pickup):
```
Comment (Justintime50) [inline tests/test_pickup.py:6]: Let's call this `create_pickup` instead, it more clearly defines what's happening.

File: tests/test_pickup.py [hunk 16]
```diff
@@ -1,148 +1,56 @@
...
-        instructions="Special pickup instructions",
-    )
-
-    assert pickup.pickup_rates != [], pickup.messages
-
-    pickup.buy(carrier=pickup.pickup_rates[0].carrier, service=pickup.pickup_rates[0].service)
+def test_pickup_buy(usps, one_call_buy_shipment, basic_pickup):
+    pickup = generate_pickup(one_call_buy_shipment, basic_pickup)
+
+    bought_pickup = pickup.buy(carrier=usps, service="Nextday")
```
Comment (Justintime50) [inline tests/test_pickup.py:38]: This should be `NextDay` (capital D) and could be a fixture just like the `usps_service` is (call it `pickup_service`)

File: tests/test_shipment.py [hunk 17]
```diff
@@ -1,228 +1,157 @@
...
     assert smartrates[0]["time_in_transit"]["percentile_85"] is not None
     assert smartrates[0]["time_in_transit"]["percentile_90"] is not None
     assert smartrates[0]["time_in_transit"]["percentile_95"] is not None
     assert smartrates[0]["time_in_transit"]["percentile_97"] is not None
     assert smartrates[0]["time_in_transit"]["percentile_99"] is not None
+
+
+@pytest.mark.vcr()
+def test_shipment_create_empty_objects(basic_shipment):
+    shipment_data = basic_shipment
```
Comment (Justintime50) [inline tests/test_shipment.py:116]: You also need to create this shipment with an empty list for `customs_info` to test we don't break things with empty lists. You'll then need to assert that.

File: tests/test_shipment.py [hunk 18]
```diff
@@ -1,228 +1,157 @@
...
+
+
+@pytest.mark.vcr()
+def test_shipment_create_with_ids(basic_address, basic_parcel):
+    from_address = easypost.Address.create(**basic_address)
+    to_address = easypost.Address.create(**basic_address)
+    parcel = easypost.Parcel.create(**basic_parcel)
+
+    shipment = easypost.Shipment.create(
+        from_address={"id": from_address.id}, to_address={"id": to_address.id}, parcel={"id": parcel.id}
```
Comment (Justintime50) [inline tests/test_shipment.py:149]: Please always use trailing commas, Black will format this accordingly when you do so (breaking these up to their own lines for better readability).

```suggestion
        from_address={"id": from_address.id}, to_address={"id": to_address.id}, parcel={"id": parcel.id},
```

File: tests/test_tracker.py [hunk 19]
```diff
@@ -1,85 +1,46 @@
...
-    assert trackers2["has_more"] is not None
-    assert trackers2["trackers"][0].id == tracker3.id  # Should be the same as the id for tracker3
+    assert len(trackers_array) <= page_size
+    assert trackers["has_more"] is not None
+    assert all(isinstance(tracker, easypost.Tracker) for tracker in trackers_array)
 
 
 @pytest.mark.vcr()
 def test_tracker_create_list():
-    """Tests that we can create a list of trackers in bulk."""
```
Comment (Justintime50) [inline tests/test_tracker.py:74]: This comment was valid and should be left.

File: tests/test_tracker.py [hunk 20]
```diff
@@ -1,85 +1,46 @@
...
-    success = easypost.Tracker.create_list(
+    response = easypost.Tracker.create_list(
         {
             "0": {"tracking_code": "EZ1000000001"},
-            "1": {"tracking_code": "EZ2000000002"},
-            "2": {"tracking_code": "EZ3000000003"},
-            "3": {"tracking_code": "EZ4000000004"},
```
Comment (Justintime50) [inline tests/test_tracker.py:80]: This line was valid and could be left.

File: tests/test_tracker.py [hunk 21]
```diff
@@ -1,85 +1,46 @@
...
+            "1": {"tracking_code": "EZ1000000002"},
+            "2": {"tracking_code": "EZ1000000003"},
         }
     )
 
-    # This endpoint/method does not return anything, just make sure the request doesn't fail
```
Comment (Justintime50) [inline tests/test_tracker.py:84]: This comment was valid and should be left.

File: easypost/easypost_object.py [hunk 22]
```diff
@@ -183,6 +183,9 @@ def __repr__(self):
     def __str__(self):
         return self.to_json(indent=2)
 
+    def __eq__(self, other):
```
Comment (nwithan8) [inline easypost/easypost_object.py:186]: I'd recommend putting a check here, like 
```python
if not isinstance(other, EasyPostObject):
     return False
````
to short-circuit the function if we're not comparing a non-EasyPostObject to this clearly EasyPostObject.

File: tests/test_user.py [hunk 23]
```diff
@@ -1,45 +1,81 @@
...
+    assert str.startswith(user.id, "user_")
+    assert user.phone == test_phone
 
 
 @pytest.mark.vcr()
-def test_child_user_create(prod_api_key):
-    # Create an address and then verify some fields to test whether it was created just fine.
-    child_user = easypost.User.create(name="Python All-Things-Testing")
+@pytest.mark.skip(
+    reason="Due to our inability to create child users securely,"
```
Comment (Justintime50) [inline tests/test_user.py:48]: You'll need a space after the comma or before the next `we`, when concatenating strings it'll slap them together without a space unless you add one yourself.

File: tests/test_webhook.py [hunk 24]
```diff
@@ -1,42 +1,52 @@
...
-    expected_url = "http://example.com/{0}".format(per_run_unique)
-
-    # Create a webhook
-    webhook = easypost.Webhook.create(url=url)
-    assert webhook.id is not None
-    assert webhook.mode == "test"
-    assert webhook.url == expected_url
-    assert webhook.disabled_at is None
+def test_webhook_create():
+    webhook = easypost.Webhook.create(url="http://example.com")
```
Comment (Justintime50) [inline tests/test_webhook.py:8]: Nit: You could make the webhook url a fixture like the others so you can reuse it and if ever we needed to switch it it's easier.

**Review by Justintime50 — CHANGES_REQUESTED** [review]
Once all the below has been addressed, I'll be good to stamp. Looked through the whole thing.

Please also remove the `pytz` dev dependency and associated TODO since it's no longer needed with this overhaul.

File: tests/test_order.py [hunk 25]
```diff
@@ -1,95 +1,52 @@
...
+
+
+@pytest.mark.vcr()
+def test_order_retrieve(basic_address, basic_shipment):
+    order = create_order(basic_address, basic_shipment)
+
+    retrieved_order = easypost.Order.retrieve(order.id)
+
+    assert isinstance(retrieved_order, easypost.Order)
+    assert retrieved_order.id == order.id
```
Comment (nwithan8) [inline tests/test_order.py:30]: Add a note to the `retrieve` test in `test_order`, `test_insurance`, `test_batch` and `test_tracker`:
```
# status changes between creation and retrieval, so we can't compare the whole object
```

**Review by nwithan8 — CHANGES_REQUESTED** [review]

File: tests/test_order.py (hunk 25 above)
Comment (Justintime50) [reply tests/test_order.py:30]: 100% agree and thought to say the same, but didn't as we didn't do this in the other libs (but should have).

File: tests/test_address.py [hunk 26]
```diff
@@ -1,144 +1,81 @@
...
-
-    assert len(exception["error"]["errors"]) >= 2
-    assert exception["error"]["errors"][0]["message"] == "Address not found"
-    assert exception["error"]["errors"][1]["message"] == "House number is missing"
+def test_address_create_verify(incorrect_address_to_verify):
+    """
+    Test creating a verified address.
+    We purposefully pass in slightly incorrect data to get the corrected address back once verified.
+    """
+    address = easypost.Address.create(**incorrect_address_to_verify)
```
[the remark this answers is not in the document]
Comment (jchen293) [reply tests/test_address.py:54]: I will remove this function since the testing duplicate to the one you pointed out in the above comment: https://github.com/EasyPost/easypost-python/pull/162#discussion_r829439481

File: tests/test_carrier_account.py [hunk 27]
```diff
@@ -0,0 +1,75 @@
+import pytest
+
+import easypost
+
+
+def generate_carrier_account():
```
[the remark this answers is not in the document]
Comment (jchen293) [reply tests/test_carrier_account.py:6]: We had this discussion yesterday. Since this helper function is only being used in `test_carrier_account.py`, we should be putting it in `confest.py`

[the remark this answers is not in the document]
Comment (Justintime50) [reply tests/test_carrier_account.py:6]: I misunderstood what you were asking. Because this is `static data`, it belongs in `conftest.py` along with the other fixtures of static data. If this function actually did something (aside from returning a dictionary) and was only used here; then yes, having this helper function in this file would make sense but that's not the case here.

File: tests/test_shipment.py (hunk 17 above)
Comment (jchen293) [reply tests/test_shipment.py:116]: The assertions are correct and passed, I didn't see why we need it since it's already null. Anyway, I will add this back.

File: tests/test_tracker.py (hunk 20 above)
Comment (jchen293) [reply tests/test_tracker.py:80]: I was using the test input example from other libs to make it consistent, the ` "3": {"tracking_code": "EZ4000000004"},` is unnecessary, let's keep what I have now.

File: tests/test_user.py (hunk 23 above)
Comment (jchen293) [reply tests/test_user.py:48]: thanks for letting me know that!

File: tests/test_carrier_account.py [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+        },
+    }
+
+
+@pytest.mark.vcr()
+def test_carrier_account_create(prod_api_key):
+    carrier_account = easypost.CarrierAccount.create(**generate_carrier_account())
+
+    assert isinstance(carrier_account, easypost.CarrierAccount)
+    assert str.startswith(carrier_account.id, "ca_")
```
[the remark this answers is not in the document]
Comment (Justintime50) [reply tests/test_carrier_account.py:23]: Nah, there is a `type` property on this object and it should be `UpsAccount`.

File: tests/test_shipment.py (hunk 17 above)
Comment (Justintime50) [reply tests/test_shipment.py:116]: I understand these assertions are correct. I'm asking that **in addition** to these assertions you already have, you should also pass an empty list to `customs_info`. This tests purpose is to check that we encode data to the API correctly. We had big problems with the PHP encoding because we didn't test enough edge cases (users frequently pass us empty lists) and so we should test/guard against that by having a test that does this as we do in all our other libs.

Comment (jchen293) [reply tests/test_shipment.py:116]: Thanks for the explanation, I have added an empty list to the customs_info field.

File: tests/test_carrier_account.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (jchen293) [reply tests/test_carrier_account.py:23]: Make sense, I was following the PHP test suite. I will add this assertion.

File: tests/test_batch.py [hunk 29]
```diff
@@ -1,92 +1,106 @@
...
+
+@pytest.mark.vcr()
+def test_batch_retrieve(one_call_buy_shipment):
+    batch = easypost.Batch.create(shipments=[one_call_buy_shipment])
+
+    retrieved_batch = easypost.Batch.retrieve(batch.id)
+
+    assert isinstance(batch, easypost.Batch)
+    assert retrieved_batch.id == batch.id
+    # status changes between creation and retrieval, so we can't compare the whole object
```
Comment (Justintime50) [inline tests/test_batch.py:23]: A comment after the code its referencing is confusing. Please always have comments either come right before the line they are referencing or on the same line (depending on length and context.)

File: tests/test_carrier_account.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (Justintime50) [reply tests/test_carrier_account.py:11]: Yeah, as it was the first one done, there is a thing here and there that were only discovered later.

File: tests/test_carrier_account.py [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+
+@pytest.mark.vcr()
+def test_carrier_account_retrieve(prod_api_key, basic_carrier_account):
+    carrier_account = easypost.CarrierAccount.create(**basic_carrier_account)
+
+    retrieved_carrier_account = easypost.CarrierAccount.retrieve(carrier_account.id)
+
+    assert isinstance(retrieved_carrier_account, easypost.CarrierAccount)
+    assert carrier_account.id == retrieved_carrier_account.id
+    # status changes between creation and retrieval, so we can't compare the whole object
```
Comment (Justintime50) [inline tests/test_carrier_account.py:25]: Comment before the line it's referencing.

File: tests/test_easypost.py [hunk 31]
```diff
@@ -0,0 +1,11 @@
...
+
+
+def test_timeout():
+    assert requestor.timeout == 60
+
+    requestor.timeout = 1
+
+    assert requestor.timeout == 1
+
+    requestor.timeout = 60
```
[the remark this answers is not in the document]
Comment (Justintime50) [reply tests/test_easypost.py]: No comment was added here but it was marked resolved. Please only mark items resolved if they actually get resolved.

File: tests/test_insurance.py [hunk 32]
```diff
@@ -1,64 +1,44 @@
...
+    insurance_data = basic_insurance
+    insurance_data["tracking_code"] = shipment.tracking_code
+
+    insurance = easypost.Insurance.create(**insurance_data)
+
+    retrieved_insurance = easypost.Insurance.retrieve(insurance.id)
+
+    assert isinstance(retrieved_insurance, easypost.Insurance)
+    assert insurance.id == retrieved_insurance.id
+    # status changes between creation and retrieval, so we can't compare the whole object
```
Comment (Justintime50) [inline tests/test_insurance.py:33]: Comment before code it's referencing.

File: tests/test_order.py [hunk 33]
```diff
@@ -1,95 +1,53 @@
...
+    # status changes between creation and retrieval, so we can't compare the whole object
```
Comment (Justintime50) [inline tests/test_order.py:31]: Comment before code it's referencing.

File: tests/test_shipment.py [hunk 34]
```diff
@@ -1,228 +1,160 @@
...
+    shipment_data["options"] = None
+    shipment_data["tax_identifiers"] = None
+    shipment_data["reference"] = ""
+
+    shipment = easypost.Shipment.create(**shipment_data)
+
+    assert isinstance(shipment, easypost.Shipment)
+    assert str.startswith(shipment.id, "shp_")
+    assert shipment.options is not None
+    assert shipment.customs_info is None
```
Comment (Justintime50) [inline tests/test_shipment.py:127]: I apologize if I mispoke here. The test should be an empty `customs_items` nested list (see PHP here: https://github.com/EasyPost/easypost-php/blob/master/test/EasyPost/ShipmentTest.php#L237).

We should then assert against the top level `tax_identifiers` (missing) and assert the nested `customs_items`. This way, if we ever changed encoding in this lib, it'd catch the regression because we are testing all sorts of things (empty strings, empty objects - top level and nested, objects that are `None`, etc)

File: tests/test_tracker.py [hunk 35]
```diff
@@ -1,85 +1,49 @@
...
+    tracker = easypost.Tracker.create(tracking_code="EZ1000000001")
 
-    assert tracker2.id == tracker.id  # Should be the same as above
+    retrieved_tracker = easypost.Tracker.retrieve(tracker.id)
 
-    # retrieve all trackers by tracking_code
-    trackers = easypost.Tracker.all(tracking_code=tracking_code)
+    assert isinstance(retrieved_tracker, easypost.Tracker)
+    assert retrieved_tracker.id == tracker.id
+    # status changes between creation and retrieval, so we can't compare the whole object
```
Comment (Justintime50) [inline tests/test_tracker.py:23]: Comment before code referencing it.

**Review by Justintime50 — CHANGES_REQUESTED** [review]
Really close, a couple small last things.

File: tests/test_shipment.py (hunk 34 above)
Comment (jchen293) [reply tests/test_shipment.py:127]: My assertions are correct as well as the order, do you mean assert `tax_identifiers` is null which I don't see other libs have this yet.

Comment (jchen293) [reply tests/test_shipment.py:127]: We will need to patch node.js lib as well for this one: https://github.com/EasyPost/easypost-node/blob/master/test/resources/shipment.test.js#L34

**Review by nwithan8 — APPROVED** [review]

**Comment by nwithan8** [comment]
LGTM, thanks for working so hard on this!

File: tests/test_shipment.py (hunk 34 above)
Comment (Justintime50) [reply tests/test_shipment.py:128]: Yes please, add the assertion requested here.

Comment (jchen293) [reply tests/test_shipment.py:128]: I will create a ticket in the backlog to add the assertion in all 7 libs so I don't dismiss Nate's approval again.

**Review by Justintime50 — APPROVED** [review]
Nice work!
